dimscord/objects.nim

```diff
@@ -37,6 +37,7 @@ proc newShard*(id: int, client: DiscordClient): Shard =
         ),
         retry_info: (ms: 1000, attempts: 0)
     )
+var client*: DiscordClient
```

Comment:
While people don't usually run multiple bots, having a global like this is asking for trouble.

Maybe a reference to the API client inside Message/Interaction/etc objects would be better? Would add a small overhead to each object but don't think it'd be noticable when network is the most costly operation

Reply:
Personally I dont think adding a _client field to some object like Message, since it just feels out of place and not fitting also if its in other objects that are not `ref object` for a reason then _client would not be much up to date which could issues like with ratelimit handling where if the route is processing or ratelimited etc.

Reply:
`DiscordClient` is a `ref` so they'll all be pointing to the same object

Reply:
> While people don't usually run multiple bots, having a global like this is asking for trouble.
> 
> Maybe a reference to the API client inside Message/Interaction/etc objects would be better? Would add a small overhead to each object but don't think it'd be noticable when network is the most costly operation

Ah, nice catch ! Haven't thought about multibotting (or do they say sharding in discord-land ?) this can be very troubling indeed... I really don't know what we can do about this though other than what has already been said.


> Personally I dont think adding a _client field to some object like Message is a good idea either, since it just feels out of place and not fitting

Well, atm I cannot say if this is going to be the solution to this problem anyways, I'll experiment and see

Reply:
> While people don't usually run multiple bots, having a global like this is asking for trouble.

I have thought about what potential trouble would arise from this, first of all let's keep in mind some facts:
- We're solely using `client` as a read-only ref after it has been set
- `Shard` and `DiscordClient` are circular refs and every `Shard.client` is pointing back to the same `DiscordClient` 
- The library takes cares of sharding for people who may want to without using threads and whatnot

Those may seem like they're not relevant but here's the thing:
In my pov, since we only care about `DiscordClient` and every `Shard` refers to the same object and have the same informations, I believe that the "multiple bot" issue isn't actually a big deal but maybe you thought of something different ?

### So, with that in mind, here is what i propose:
```nim
proc getClient*(c = none DiscordClient): DiscordClient =
    ## Get the current state of `DiscordClient`
    
    # as a proc, we can use pragmas to make usage of `DiscordClient` safe and futureproof.
    # `c = none DiscordClient` is a default value that allows us to not specify an argument when we just get the object 
    # {.global.} pragma means `private_client` can live past the execution of `getClient()`

    # `private_client` is guarded by `getClient`'s scope, we cannot access it from the outside nor mutate it.
    var private_client {.global.}: DiscordClient

    # `once` template allows means to not assign a ref to `private_client` more than twice (could use `let` for that though)
    once:
        private_client = get(c)
    
    result = private_client
```
### inside `startSession`:
```nim
discard getClient(some(discord))
```